### PR TITLE
Migrate the Microsoft Android SDK with Microsoft Java SDK

### PR DESCRIPTION
**Pull request by vikyding**
Update Readme file for instruction on how to import and use Microsoft Java SDK
Update Demos step 4 for integration with Microsoft Graph 
Update MSGraphServiceClientManager and MSGraphServiceController classes'methods to call Microsft Graph Java SDK functions

@jthake-msft I closed all previous partially PR and create this for the whole projects, please kindly review it. Thanks.

### CODE REVIEW

**Comment by jthake-msft** [comment]
@elisenyang will also be reviewing this.

File: Demos/04-add-msgraph/README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Integrate Microsoft Graph into the Application
 
-The last step is to incorporate the Microsoft Graph into the application. For this application, you will use the Microsoft Graph Android SDK.
+The last step is to incorporate the Microsoft Graph into the application. For this application, you will use the Microsoft Graph Java SDK.
```
Comment (jasonjoh) [inline Demos/04-add-msgraph/README.md]: This file still calls it Android SDK in other places. Make sure it's updated everywhere

File: Demos/04-add-msgraph/README.md [hunk 2]
```diff
@@ -9,10 +9,17 @@ The last step is to incorporate the Microsoft Graph into the application. For th
     1. Add the following code to the `dependencies` section, immediately after the previously added dependencies:
 
         ```gradle
-        implementation 'com.microsoft.graph:msgraph-sdk-android:1.3.2'
+        implementation 'com.microsoft.graph:microsoft-graph:0.1.+'
         ```
 
     1. Sync the dependencies with the project by selecting **File > Sync Project with Gradle Files**.
+        
+        if cannot reslove the package, add the following code to the project build.gradle file 'allprojects' section:
```
Comment (jasonjoh) [inline Demos/04-add-msgraph/README.md]: Put this in a callout (with the `>` character). Also wrap the code in a fenced code block.

File: Demos/04-add-msgraph/build.gradle [hunk 3]
```diff
@@ -19,6 +19,10 @@ allprojects {
     repositories {
         google()
         jcenter()
+        mavenCentral()
```
Comment (jasonjoh) [inline Demos/04-add-msgraph/build.gradle]: is `mavenCentral()` needed? I omitted it and it worked ok for me.

File: Demos/04-add-msgraph/README.md [hunk 4]
```diff
@@ -31,8 +38,8 @@ The last step is to incorporate the Microsoft Graph into the application. For th
         import com.microsoft.graph.authentication.IAuthenticationProvider;
         import com.microsoft.graph.core.DefaultClientConfig;
         import com.microsoft.graph.core.IClientConfig;
-        import com.microsoft.graph.extensions.GraphServiceClient;
-        import com.microsoft.graph.extensions.IGraphServiceClient;
+        import com.microsoft.graph.requries.extensions.GraphServiceClient;
```
Comment (jasonjoh) [inline Demos/04-add-msgraph/README.md]: Typo in import: requries

File: Demos/04-add-msgraph/app/build.gradle [hunk 5]
```diff
@@ -16,6 +16,10 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
         }
     }
+
+    packagingOptions{
+        pickFirst 'META-INF/jersey-module-version'
```
Comment (jasonjoh) [inline Demos/04-add-msgraph/app/build.gradle]: I didn't see this in the module 4 README, but I had to add this line to get the project to run after doing the final steps in the README. Should this get added?

**Comment by jasonjoh** [comment]
@vikyding The changes that you make to the module 4 README have to also be incorporated into Lab.md

File: Demos/04-add-msgraph/README.md (hunk 1 above)
Comment (vikyding) [reply Demos/04-add-msgraph/README.md]: Thanks for suggestions. I updated all other places and also made changes for all following review comments. Please kindly review it again. Thanks.

File: Demos/04-add-msgraph/README.md (hunk 2 above)
Comment (vikyding) [reply Demos/04-add-msgraph/README.md]: Updated it. Thanks.

File: Demos/04-add-msgraph/app/build.gradle (hunk 5 above)
Comment (vikyding) [reply Demos/04-add-msgraph/app/build.gradle]: Yes, added it.

**Comment by andrewconnell** [comment]
These changes => Java SDK are breaking the lab. I'm doing a refresh to the module to verify everything works as part of a regularly quarterly refresh. Per request of @jthake-msft I'm running through the lab in this PR as part of the refresh.

Unfortunately, when I run through the lab I'm getting errors trying to use the Java SDK. Therefore this refresh is blocked by this PR and needs to be addressed by the author. 

Specifically when I add the reference to `implementation 'com.microsoft.graph:microsoft-graph:0.1.+'`, it syncs just fine without errors, but the next step in creating the `MSGraphServiceClientManager` utility class doesn't resolve any of the import statements for `com.microsoft.graph.*`. 

Further, I went back and added the `maven { url "https://dl.bintray.com/microsoftgraph/Maven" }` line and sync'd again just to see if that would fix it, but this throws another error in the console:

image

If something else needs to be installed outside of the steps in order for maven to work, those steps must be added to the lab as they are not listed as prereq's for the lab.

In addition, the changes submitted by the author do not follow the official MSFT Style Guide. I can fix these in the refresh, but not until someone addresses these blocking issues.

**Comment by vikyding** [comment]
@jthake-msft @andrewconnell Sure. I'll take a look and figure it.

**Comment by andrewconnell** [comment]
@vikyding do you by chance have an ETA when you think this will be ready for me to take a look again to perform the refresh for this quarter? 

I have a bunch of modules to refresh and because of this issue, I need to reschedule this one to make it gets addressed before Ignite in 1.5 weeks.

File: Demos/04-add-msgraph/README.md [hunk 6]
```diff
@@ -90,10 +97,10 @@ The last step is to incorporate the Microsoft Graph into the application. For th
         import com.microsoft.graph.concurrency.ICallback;
         import com.microsoft.graph.core.ClientException;
         import com.microsoft.graph.extensions.Event;
```
Comment (vikyding) [inline Demos/04-add-msgraph/README.md]: update it to com.microsoft.graph.models.extensions.Event;

**Comment by vikyding** [comment]
@andrewconnell Hi Andrew, sorry for letting you wait. I recreated the demo following the steps but I didn't reproduce the same errors. So I did an investigation for your issues and found some reasons or solutions for them. 

1. Cannot resolve any of the import statements for com.microsoft.graph.* after sync Gradle file successfully.
Actually, com.microsoft.graph is not a package name so it cannot be imported directly like this. If you want to avoid multiple imports from the same package, please use import package name such as 

import com.microsoft.graph.authentication.*;
import com.microsoft.graph.core.*;
import com.microsoft.graph.http.*;
import com.microsoft.graph.requests.extensions.*;
import com.microsoft.graph.models.extensions.*;

2. Only add the "maven { url "https://dl.bintray.com/microsoftgraph/Maven" }" to the project Gradle file when your gradle cannot find "com.microsoft.graph:microsoft-graph:0.1.+" during async to specific the maven repo location. For the error cannot found maven{} in the console, it may be caused by gradle version. I tried many times and didn't reproduce the same error this time. But I did meet it before in other projects. I may edit the readme file for more specific.

3. Ther are nothing else needs to be installed outside of the steps. Since the Android Studio creates a project based on Gradle. Different Gradle version may cause some differences during sync. I found the previous steps didn't mention the Gradle version and I thought it should be decided by customers so I didn't put it in steps. 

4. There are some typos haven't been updated. So I will update them before tomorrow morning. Then you can take a look again for refresh tomorrow. 

5. Thank you so much for fixing the unofficial changes for me. I just joined the Microsoft and did not know about MSFT Style Guide. But I'll read it soon. I very appreciate your help and do the favor for you to refresh this module. Just pin me if you still have problems when refreshing. I'll pay extra attention to this repo until it refreshes successfully. Thank

**Comment by vikyding** [comment]
@andrewconnell miss * in import package

import com.microsoft.graph.authentication.*;
import com.microsoft.graph.core.*;
import com.microsoft.graph.concurrency.*;
import com.microsoft.graph.http.*;
import com.microsoft.graph.logger.*;
import com.microsoft.graph.models.extensions.*;
import com.microsoft.graph.requests.generated.*;
import com.microsoft.graph.serializer.*;

**Comment by andrewconnell** [comment]
@vikyding Thanks for taking a look and making updates. When updating the lab, please don't update the demos. the demo README files are copies from the exercises in the lab.md file in the root of the project. We typically update the lab.md file to make sure the whole things works, THEN we update each demo's readme.md file to reflect the changes from the lab.md file. The lab is what students / consumers of the module will be primarily using, not the demos.

So maven isn't something you need to install in order for it to work? Its included in the base Andriod Studio? It wasn't something I have seen in the default Andriod Studio install.

I'll take another stab at the refresh in the morning my time (UTC -0400) as I see you're pushing these updates right now... so I'll give you a chance to finish any other work you're doing.

**Comment by vikyding** [comment]
@andrewconnell Thanks for notifying me that. I updated the Lab.md file for steps, please ignore the updates for demos README files right now. And yes, Maven doesn't need to install for it to work. When created a new android project in Android Studio first time, it will automatically install Android tools for Gradle. So there is nothing else need to install for this project. And I have finished my updates and you can have a refresh for this one following the Lab.md to double check if there anything that I missed. Thanks.

**Comment by jthake-msft** [comment]
I'm going to close this PR, as @andrewconnell  has submitted #7 which fixes some of the issues he had. There is one outstanding issue with MSAL that i'm chasing up.

**Pull request closed without merging**